datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -1331,7 +1331,17 @@ pub fn validate_unique_names<'a>(
     })
 }
 
-/// Union two logical plans.
+/// Union two [`LogicalPlan`]s.
+///
+/// Constructs the UNION plan, but does not perform type-coercion. Therefore the
+/// subtree expressions will not be properly typed until the optimizer pass.
```

Comment:
hm, does that mean if there is no optimizer phase the query will crash?

Reply:
I think maybe this depends on how users construct their own logical plans? And if they are already ensuring type correctness in their own code?

Our logical plan construction relied upon the [previous behavior](https://github.com/influxdata/arrow-datafusion/commit/afa23abc46059e100061295b619b7b66fbc39625) of the `union()` api, which was immediate type coercion. IMO - this was not an explicit api contract, rather an implementation detail. Regardless, we use the typed-coerced union schema & expressions when adding other logical nodes (e.g. limit, sort, group by). Once the behavior of the api changed, we started building bad plans -- which produced incorrect results and in one case errored.

I filed this ticket as a doc enhancement, not a bug, since I didn't think(?) type coercion was part of the api contract. My hope was to (a) make it clear when/how unions can be type coerced, and (b) make public the APIs to do so. In other words, make it easier for users to ensure type correctness without the optimizer pass.

Reply:
Should I have phrased the docs differently? Or is there another approach we should take?

Reply:
Also, to clarify. The "incorrect plans" is because the new `union()` behavior [takes the left node's schema](https://github.com/apache/datafusion/blob/4fd460d7a2d2b60e6f9dc716148d8596d94d048c/datafusion/expr/src/logical_plan/builder.rs#L1346-L1352) -- and doesn't type cast the expressions. When we built logical plans with union + gap filling (adding casted scalar nulls), we started having missing fields etc when "reading the union" before constructing the next node.

Maybe that part should be a bug?